NEW: Improve performance when reading control values

andrew-oc opened the pull request:
### Description

This commit introduces the concept of control value caching where the system marks controls that have been changed as stale and when reading control values, either through InputControl.Readvalue or InputAction.ReadValue, only stale controls read from the underlying storage and process the result before returning it. Unchanged controls simply return the cached value to the caller.

### Changes made

The InputDeviceBuilder now creates a binary tree of ControlBitRangeNode elements to store the bit ranges that controls map to in the state memory of the device. The tree is stored in the device itself and when state is changed, either through a state event or through InputState.Change, it is traversed to figure out what controls have been changed. Changed controls are marked as stale and when reading the value of the control, only stale controls go to the device storage and pull out the value and run it through the control's processor stack. 

### Notes

The InputControl<TValue>.value property is ref readonly to improve performance when using some of the larger input control state structs, especially ones from XR like Bone and Eyes. This necessitated a change to those structs where all the properties have been made into fields so the readonly aspect can be enforced by the compiler at the call site, so now they're straight POCOs. 

This PR also lays the ground work for some further optimizations, namely:

- removal of state monitors
  State monitors exist to watch sections of memory and fire off input action processing when those sections change. They are implemented in O(n) time as a straight linear walk through all monitors which can get expensive for fully loaded action maps. With this PR, we can now tell what controls have changed given a piece of new state memory, so state monitors might not be necessary.
- improved performance of any code that uses the InputControlEnumerator, such as rebinding, InputSystem.onEvent callbacks, and especially listening for unpaired device activitiy.
- makes it possible to correctly fix wasPressed/ReleasedThisFrame

### Checklist

Before review:

- [x] Changelog entry added.
    - Explains the change in `Changed`, `Fixed`, `Added` sections.
    - For API change contains an example snippet and/or migration example.
    - FogBugz ticket attached, example `([case %number%](https://issuetracker.unity3d.com/issues/...))`.
    - FogBugz is marked as "Resolved" with *next* release version correctly set.
- [x] Tests added/changed, if applicable.
    - Functional tests `Area_CanDoX`, `Area_CanDoX_EvenIfYIsTheCase`, `Area_WhenIDoX_AndYHappens_ThisIsTheResult`.
    - Performance tests.
    - Integration tests.
- [x] Docs for new/changed API's.
    - Xmldoc cross references are set correctly.
    - Added explanation how the API works.
    - Usage code examples added.
    - The manual is updated, if needed.

During merge:

- [ ] Commit message for squash-merge is prefixed with one of the list:
    - `NEW: ___`.
    - `FIX: ___`.
    - `DOCS: ___`.
    - `CHANGE: ___`.
    - `RELEASE: 1.1.0-preview.3`.

andrew-oc committed:
CHANGE: Improve performance of reading control values

This commit introduces the concept of control value caching where the system marks controls that have been changed as stale and when reading control values, either through InputControl.Readvalue or InputAction.ReadValue, only stale controls read from the underlying storage and process the result before returning it. Unchanged controls simply return the cached value to the caller.

andrew-oc committed:
Fix for API verification failure

andrew-oc committed:
Fixes for earlier versions of Unity

* Removed usages of MemoryMarshal and Span<T>
* Removed usages of MathF

andrew-oc committed:
Formatter changes

ekcoh commented on Packages/com.unity.inputsystem/InputSystem/Controls/InputControlExtensions.cs:
```diff
@@ -1896,6 +1897,26 @@ public DeviceBuilder WithStateOffsetToControlIndexMap(uint[] map)
+	                for (var i = 0; i < numNodes; i++)
+	                {
+		                device.m_ControlTreeNodes[i] = *(InputDevice.ControlBitRangeNode*)(nodePtr + i * sizeOfNode);
+		                // device.m_ControlTreeNodes[i] = MemoryMarshal.Read<InputDevice.ControlBitRangeNode>(
```
Intentionally uncommented code or should it be removed?

andrew-oc replied:
Should have been removed.

ekcoh commented on Packages/com.unity.inputsystem/InputSystem/Devices/InputDevice.cs:
```diff
@@ -708,6 +708,41 @@ internal bool disabledWhileInBackground
+            // us where to start for each node and how many controls this node points at. This is an unsigned short so that
+            // we could in theory support devices with up to 65553 controls.
+            public ushort controlStartIndex;
+            public byte controlCount;
```
Is there a good reason to keep controlCount as a byte here since it will anyway get padded to the size of an ushort?

andrew-oc replied:
I was going for the smallest possible size here to keep the entire tree of ranges in L1, but I forgot to add the Pack attribute! Good catch.

ekcoh reviewed (approved):
Code looks good but cannot say I follow on the design/reasoning behind the ControlBitRangeNode. I guess the ControlBitRangeNode is indirectly tested by code using it, but might be beneficial to test ControlBitRangeNode in isolation as well with some unit tests? Might also make it easier to grasp what is going on with the bits there.

ekcoh commented:
Out of curiosity, what performance improvements have you observed with relation to these changes? Based off the description it sound like this is an enabler for further improvements.

andrew-oc committed:
Minor changes for API validation suite.

andrew-oc committed:
Addressing review comments and fixing API validator issue

* Removed unused parameter from WriteChangedControlStatesInternal
* InputDevice.WriteChangedControlStatesInternal now performs an additional check at the control level to make sure control values have really definitely changed before marking them as stale to make reading state more efficient

andrew-oc commented:
> Out of curiosity, what performance improvements have you observed with relation to these changes? Based off the description it sound like this is an enabler for further improvements.

Micro-benchmarks showed about a 2x performance improvement when reading the value of simple controls with no processors, and that gets better the more complex a control is i.e. more child controls, especially AxisControl, and better again the more processors a control has. One real world scenario that should give us a solid performance improvement from this PR alone though is having something like a four way composite attached to 16 different controls (something that's given us problems before, particularly on Switch). In the past, we'd read the value of each of those 16 controls at least twice per composite ReadValue call (once for the value and once for the magnitude check). Now we'll only read the value of the composite parts that have actually changed.

andrew-oc committed:
Add API validation exception for issue with ref returns

Rene-Damm commented:
Beautiful work.

andrew-oc committed:
Aaaand removing validation exceptions again...

They have absolute paths in them so they don't match on CI. Have to wait for the API validator fixes regarding ref returns in overridden members before validation will pass.

andrew-oc commented:
> Beautiful work.

:bow:

jimon commented on Packages/com.unity.inputsystem/InputSystem/Devices/InputDevice.cs:
```diff
@@ -708,6 +709,55 @@ internal bool disabledWhileInBackground
+        // When a device gets built from a layout, we create a binary tree from its controls where each node in the tree
+        // represents the range of bits that cover the left or right section of the parent range. For example, starting
+        // with the entire device state block as the parent, where the state block is 100 bits long, the left node will
+        // cover from bits 0-50, and the right from bits 50-100. For the left node, we'll get two more child nodes where
```
nitpick: 0-50 and 51-99, because in 100 bit log state block bit indices are from 0 to 99 inclusive

jimon commented on Packages/com.unity.inputsystem/InputSystem/Devices/InputDevice.cs:
```diff
@@ -906,5 +956,130 @@ internal static TDevice Build<TDevice>(string layoutName = default, string layou
+
+        private static unsafe bool HasDataChangedInRange(byte* deviceStatePtr, void* statePtr, uint startOffset, uint sizeInBits)
+        {
+            if (sizeInBits == 1)
```
Is this branch necessary? Feels like MemCmp should contain single bit optimization instead

andrew-oc replied:
It should but it has too many arguments that also need to be checked before you can decide to do the single bit optimization and you end up with more branches instead of fewer.

jimon commented on Packages/com.unity.inputsystem/InputSystem/Controls/InputControl.cs:
```diff
@@ -991,17 +1022,73 @@ internal virtual IEnumerable<object> GetProcessors()
+        /// <remarks>
+        /// Note that this property implements caching to avoid applying processors when the underlying
+        /// control has not changed. With this in mind, be aware of processors that use global state, such
+        /// as the <see cref="Processors.AxisDeadzoneProcessor"/>. Unless the control has been actuated, the processors
```
Feels like we would benefit from a diagnostic mode where we always run full value read from with-in value and compare it to cached value, that way we would catch situations like changing globals in processors and make it easier to diagnose them

andrew-oc replied:
Agree and I started writing this but then it got out of hand. Will add it as a separate PR at some point

jimon reviewed (approved):
LGTM, nice work 👍 
I have some feeling that maybe there are more places where `MarkAsStale` should be called, and maybe we forgot some, so maybe it makes sense to add an escape hatch just-in-case, e.g. some internal feature flag that always forces use of uncached values.

jimon reviewed (changes requested):
Would probably be a better idea to add an opt-in or opt-out flag to how useCachedValue is computed so we are able to hot-fix it by disabling the feature if it breaks in the wild

andrew-oc committed:
Put control value caching behind a conditional flag

* The flag is not enabled by default, as this change has some behavioural changes.
* Also re-added the properties to Eyes and Bone that were removed previously as that would have been a breaking change for any assemblies that were compiled against the previous version.
* Regenerated Fast* device layouts due to an off-by-one error in the control tree building logic

andrew-oc committed:
Merge branch 'develop' into andrew/write-first-performance-optimization

jamesmcgill commented on Assets/Tests/InputSystem/APIVerificationTests.cs:
```diff
@@ -517,22 +537,27 @@ private static string FilterIgnoredChanges(string line)
 
+    /// <summary>
+    /// Use a scoped exclusion property to exclude members of a type from API verification when the member's names are not
+    /// unique in the entire project and you don't want to exclude the unrelated members. This type will scope the exlusion
```
Teeny-tiny typo "exlusion"

jamesmcgill commented on ProjectSettings/ProjectSettings.asset:
```diff
@@ -621,7 +621,7 @@ PlayerSettings:
   webGLWasmStreaming: 0
   scriptingDefineSymbols:
-    1: UNITY_ENABLE_STEAM_CONTROLLER_SUPPORT
+    1: UNITY_ENABLE_STEAM_CONTROLLER_SUPPORT;UNITY_INPUT_SYSTEM_CONTROL_VALUE_CACHING
```
Intentional I assume.
Would this mean CI and/or local TestRunner tests would have this enabled?
What about packaged Sample projects. Would they be affected?
Any other implications?

jamesmcgill reviewed (approved):
Nice! +1000 bonus points for fixing WasPressedThisFrame

ekcoh commented on Packages/com.unity.inputsystem/InputSystem/Controls/InputControl.cs:
```diff
@@ -991,17 +1022,77 @@ internal virtual IEnumerable<object> GetProcessors()
+        {
+            get
+            {
+#if UNITY_INPUT_SYSTEM_CONTROL_VALUE_CACHING
```
New API so not guarding for existing behavior

jimon commented on Packages/com.unity.inputsystem/InputSystem/Controls/InputControl.cs:
```diff
@@ -991,17 +1022,77 @@ internal virtual IEnumerable<object> GetProcessors()
+        /// <seealso cref="value"/>
         public TValue ReadValue()
         {
             unsafe
```
Let's get ReadValue optimized as well

Commit:
Merge branch 'develop' into andrew/write-first-performance-optimization

Commit:
Unified caching for ReadValue and value, refactored to use runtime setting, added caching policy to processors.

Commit:
Place back backing fields

andrew-oc committed:
Merge branch 'develop' into andrew/write-first-performance-optimization

# Conflicts:
#	Packages/com.unity.inputsystem/CHANGELOG.md

andrew-oc committed:
Minor documentation changes and changed read value caching to be off by default

andrew-oc committed:
Put back API verification excemption for fields added to Bone and Eyes

Rene-Damm commented on Packages/com.unity.inputsystem/InputSystem/Controls/InputProcessor.cs:
```diff
@@ -76,6 +76,27 @@ internal static Type GetValueTypeFromType(Type processorType)
+        }
+
+        /// <summary>
+        /// Caching policy of the processor. Override this property to provide a different value.
```
Should mention that `CacheResult` is the default.
